src/WebJobs.Script/Binding/Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -0,0 +1,320 @@
+                {
+                    // if we're binding to a user Type, we'll have a contract,
+                    // otherwise none
+                    return _bindingDataProvider != null ? _bindingDataProvider.Contract : null;
+                }
+            }
+
+            public Type TriggerValueType
+            {
+                get { return typeof(HttpRequestMessage); }
+            }
+
+            public async Task<ITriggerData> BindAsync(object value, ValueBindingContext context)
+            {
+                HttpRequestMessage request = value as HttpRequestMessage;
+                if (request == null)
+                {
+                    throw new NotSupportedException("An HttpRequestMessage is required");
+                }
+
+                IValueProvider valueProvider = null;
+                IReadOnlyDictionary<string, object> bindingData = null;
+                string invokeString = request.ToString();
+                if (_isUserTypeBinding)
+                {
+                    valueProvider = await CreateUserTypeValueProvider(request, invokeString);
+                    if (_bindingDataProvider != null)
+                    {
+                        // binding data is defined by the user type
+                        // the provider might be null if the Type is invalid, or if the Type
+                        // has no public properties to bind to
+                        bindingData = _bindingDataProvider.GetBindingData(valueProvider.GetValue());
+                    }
+                }
+                else
+                {
+                    valueProvider = new HttpRequestValueBinder(_parameter, request, invokeString);
+                    bindingData = GetRequestBindingData(request);
+                }
+
+                return new TriggerData(valueProvider, bindingData);
+            }
+
+            public Task<IListener> CreateListenerAsync(ListenerFactoryContext context)
+            {
+                if (context == null)
+                {
+                    throw new ArgumentNullException("context");
+                }
+
+                return Task.FromResult<IListener>(new NullListener());
+            }
+
+            public ParameterDescriptor ToParameterDescriptor()
+            {
+                return new TriggerParameterDescriptor
+                {
+                    Name = _parameter.Name
+                };
+            }
+
+            private static IReadOnlyDictionary<string, object> GetRequestBindingData(HttpRequestMessage request)
+            {
+                // in non strongly typed scenarios, we'll pull binding data from
+                // the uri/body
+                Dictionary<string, object> bindingData = new Dictionary<string, object>(StringComparer.OrdinalIgnoreCase);
+                if (request.Content != null && request.Content.Headers.ContentLength > 0)
+                {
+                    string body = request.Content.ReadAsStringAsync().Result;
```

Comment:
Should we make this method `async` (the caller is an `async` method as well) and `await` here to avoid `.Result`?